lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java

```diff
@@ -42,6 +43,25 @@ public void writeData(SliceOutput sliceOutput)
         };
     }
 
+    static ParquetDataOutput createDataOutput(ChunkedSliceOutput chunkedSliceOutput)
+    {
+        requireNonNull(chunkedSliceOutput, "chunkedSliceOutput is null");
+        return new ParquetDataOutput()
+        {
+            @Override
+            public int size()
+            {
+                return chunkedSliceOutput.size();
+            }
+
+            @Override
+            public void writeData(SliceOutput sliceOutput)
```

Comment:
I wonder do we actually need a copy now that we use `ChunkedSliceOutput `? 

Could we retain buffer and adjust it adaptively to avoid extra copy?

Reply:
Anyway, this could be future improvment

Reply:
If you mean reusing the compression buffer byte array between pages, then yes I am experimenting with that separately. That only saves on new array allocations and doesn't avoid copying.

Reply:
> That only saves on new array allocations and doesn't avoid copying.

OK. I thought we can avoid copying too by decompressing pages into reusable buffer

Reply:
If the buffer is reused across multiple pages, then I think we have to copy the compressed data out as we accumulate many pages worth of data before flushing to storage.
For avoiding copy, we have to avoid reusing the buffer and be okay with wasted space per page

Reply:
> For avoiding copy, we have to avoid reusing the buffer and be okay with wasted space per page

Why such buffer can't be reused? I thought decompressed pages are short lived

Reply:
Compressed pages need to be kept in memory until the row group size threshold is met. So every column writer accumulates many compressed pages before a flush to output stream.
As for the page data before compression, that part is inside parquet-mr classes and not under our control.
In any case, it's hard to make meaningful improvement here while we are tied to using parquet-mr classes, we should be able to do better when we rewrite this fully as trino code.